### PR TITLE
Adds help message in error for invalid `impl for T` syntax

### LINKED ISSUE #56031: Poor diagnostic for `impl for T`
I accidentally typed `impl for T` instead of `impl T`, and the diagnostic ([playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=037ebcc0fd36d424aea1989f99343c92)) wasn't very helpful:
```
  |
3 | impl for T {}
  |          ^ expected `<` here
```

This appears to be because of this syntax:
```rust
impl for<'a> T
```

However it makes no sense in this case. Perhaps better diagnostics would look something like this:
```
  |
3 | impl for T {}
  |     ^ expected type
```

### PR DESCRIPTION
**Pull request by repnop**
Fixes #56031.

### CODE REVIEW

**Comment by lzutao** [comment]
Missing a ui test for this PR?

**Comment by rep-nop** [comment]
I'll add one shortly, thanks for the reminder :)

File: src/test/ui/issues/issue-56031.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+struct T;
+
+impl for T {}
+
+fn main() {}
```
Comment (lzutao) [inline src/test/ui/issues/issue-56031.rs]: Missing a newline here

**Comment by lzutao** [comment]
ping @petrochenkov

**Comment by petrochenkov** [comment]
I made some tweaks to improve recovery.

@bors r+

**Comment by rep-nop** [comment]
Great! Thanks @petrochenkov :+1:

**Pull request merged**